cpp/src/arrow/array/validate.cc

```diff
@@ -699,55 +713,173 @@ struct ValidateArrayImpl {
+      return Status::Invalid("Offset invariant failure: size for slot ", slot,
+                             " out of bounds: ", offset, " + ", size, " > offset_limit");
+    }
+  }
+
+  /// \pre basic validation has already been performed
+  template <typename offset_type>
+  Status FullyValidateOffsetsAndSizes(int64_t offset_limit) {
+    const auto* validity = data.GetValues<uint8_t>(0, 0);
+    const auto* offsets = data.GetValues<offset_type>(1);
+    const auto* sizes = data.GetValues<offset_type>(2);
+
+    int64_t slot = 0;
+    if (validity) {
+      internal::BitBlockCounter counter(validity, data.offset, data.length);
+      internal::BitBlockCount block;
+      for (int64_t i = 0; i < data.length; i += block.length) {
+        block = counter.NextWord();
+        if (block.NoneSet()) {
+          continue;
+        }
+        const bool all_set = block.AllSet();
+        for (int j = 0; j < block.length; j++) {
+          slot = i + j;
+          const bool valid = all_set || bit_util::GetBit(validity, data.offset + slot);
+          if (valid) {
+            const auto size = sizes[slot];
+            if (size > 0) {
+              const auto offset = offsets[slot];
+              if (offset < 0 || offset > offset_limit) {
+                return OutOfBoundsListViewOffset<offset_type>(slot, offset_limit);
+              }
+              if (size > offset_limit - offset) {
+                return OutOfBoundsListViewSize<offset_type>(slot, offset_limit);
+              }
+            } else if (size < 0) {
+              return OutOfBoundsListViewSize<offset_type>(slot, offset_limit);
+            }
+          }
+        }
+      }
+    } else {
+      for (; slot < data.length; slot++) {
+        const auto size = sizes[slot];
+        if (size > 0) {
+          const auto offset = offsets[slot];
+          if (offset < 0 || offset > offset_limit) {
+            return OutOfBoundsListViewOffset<offset_type>(slot, offset_limit);
+          }
+          if (size > offset_limit - offset) {
+            return OutOfBoundsListViewSize<offset_type>(slot, offset_limit);
+          }
+        } else if (size < 0) {
+          return OutOfBoundsListViewSize<offset_type>(slot, offset_limit);
+        }
+      }
+    }
+
+    return Status::OK();
+  }
+
   template <typename TypeClass>
-  Status ValidateOffsets(const TypeClass& type, int64_t offset_limit) {
+  Status ValidateOffsetsAndMaybeSizes(const TypeClass&, int64_t offset_limit) {
     using offset_type = typename TypeClass::offset_type;
+    constexpr bool is_list_view = is_list_view_type<TypeClass>::value;
 
+    const bool non_empty = data.length > 0;
     if (!IsBufferValid(1)) {
-      // For length 0, an empty offsets buffer seems accepted as a special case
-      // (ARROW-544)
-      if (data.length > 0) {
-        return Status::Invalid("Non-empty array but offsets are null");
+      // For length 0, an empty offsets buffer is accepted (ARROW-544).
```

Comment:
I don't think we want to extend this special-case to list views. This is (should be?) historical.
Same for sizes below.

Reply:
I think it's natural that this issue will affect list views the same way it has affected lists.

Reply:
But I can start being more strict and wait for it.

Reply:
Note that `Buffer::Wrap(zero_len_vector)` can lead to a `Buffer` containing `nullptr` or an allocated value depending on whether the `std::vector` has increased in capacity or not (e.g. with a `.reserve()` call).